visitor date pretix sync update

ichub opened the pull request:
## Summary
closes https://github.com/proofcarryingdata/zupass/issues/246

- Residents are now defined as anyone with a 'resident' ticket.
- Visitors are now defined as anyone who is not a Resident that has a ticket in the [Zuzalu Visitors Series](https://beta.ticketh.xyz/control/event/zuzalu/zuzalu-series/).
- Each visitor ticket is associated with a subevent in the above series. Each of these subevents has a date start and end. The Zuzalu visitor week passes have dates that don't overlap, but start and end within one second of each other, which is pretty nice.
- For visitors, the pretix sync code now gets all the visitors from the above event series, and merges each visitor's potentially multiple visitor tickets into a new column within the `pretix_participant` table called `visitor_date_ranges`, which is a postgres `JsonB` that is either `undefined`, an empty array, or an array of `DateRange`s. So a visitor with multiple visitor tickets gets merged into a single `pretix_participant` zupass-side, where their multiple tickets are represented as multiple dates in that date array
- ⚠️  IMPORTANT  ⚠️  I also implemented a new feature for the pretix sync - users that are not included in pretix but do exist in our database are now **DELETED** from the zupass users list. The reason I included this change is that I infer that the new Zuzalu policy is that all visitors should have a Visitor ticket, because the organizers don't want people overstaying their welcome.

## Screenshots
### expired visitor state
If the current time is not in any of the date ranges to which this user is permissioned to be at Zuzalu, display them as 'expired' in the passport, and don't show the QR code.

<img width="497" alt="Screenshot 2023-04-28 at 6 09 35 PM" src="https://user-images.githubusercontent.com/2636237/235199016-af844527-bc8a-4d1a-baab-41878bb1b4fa.png">

### non-expired visitor state

<img width="542" alt="Screenshot 2023-04-28 at 6 12 23 PM" src="https://user-images.githubusercontent.com/2636237/235199628-2f1fa9ad-1d62-4bb2-ab54-4921602e7bd0.png">

### expired visitor state on verify screen
Realistically, this should never come up, because we do not show a QR code for expired visitors. I include this check anyways.

<img width="412" alt="Screenshot 2023-04-28 at 6 46 39 PM" src="https://user-images.githubusercontent.com/2636237/235206283-e6907912-ffcd-4c6d-90b5-874671b00b58.png">

## Testing
I tested the following.

- user was visitor only, then added as a resident. expected behavior is that the user changes into a 'resident' passport-side. actual behavior matches:

```
[PRETIX] Updating {"email":"ivansergchub+visitor@gmail.com","order_id":"UDESF","name":"Ivan Chub","role":"visitor","residence":"","email_token":"919626","visitor_date_ranges":[{"date_to":"2023-05-06T21:59:59Z","date_from":"2023-04-29T22:00:00Z"},{"date_to":"2023-04-29T21:59:59Z","date_from":"2023-04-22T22:00:00Z"},{"date_to":"2023-04-22T21:59:59Z","date_from":"2023-04-15T22:00:00Z"}]} to {"role":"resident","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"E9SES","email_token":"","visitor_date_ranges":[]}
```

- user was both visitor, and resident. cancelled resident order. expected behavior is that the user reverts back to a visitor. actual behavior matched:

```
[PRETIX] Updating {"email":"ivansergchub+visitor@gmail.com","order_id":"UDESF","name":"Ivan Chub","role":"resident","residence":"","email_token":"919626","visitor_date_ranges":[]} to {"role":"visitor","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"UDESF","email_token":"","visitor_date_ranges":[{"date_from":"2023-04-29T22:00:00Z","date_to":"2023-05-06T21:59:59Z"},{"date_from":"2023-04-22T22:00:00Z","date_to":"2023-04-29T21:59:59Z"},{"date_from":"2023-04-15T22:00:00Z","date_to":"2023-04-22T21:59:59Z"}]}
```

- user was a visitor but not a resident. they have multiple visitor entries. i delete the one that contains the order of the current week. the expected behavior is that this user is updated, and it is reflected in the passport that this participant is an expired visitor. the actual behavior matched my expectations.

```
Updating {"email":"ivansergchub+visitor@gmail.com","order_id":"UDESF","name":"Ivan Chub","role":"visitor","residence":"","email_token":"919626","visitor_date_ranges":[{"date_to":"2023-05-06T21:59:59Z","date_from":"2023-04-29T22:00:00Z"},{"date_to":"2023-04-29T21:59:59Z","date_from":"2023-04-22T22:00:00Z"},{"date_to":"2023-04-22T21:59:59Z","date_from":"2023-04-15T22:00:00Z"}]} to {"role":"visitor","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"W9XWV","email_token":"","visitor_date_ranges":[{"date_from":"2023-04-15T22:00:00Z","date_to":"2023-04-22T21:59:59Z"}]}
```

- user had several visitor entries in pretix, and I deleted all of them. I expect this to cause the pretix user to be deleted from the database. the behavior I saw matched my expectations. On the client side, this caused the user's passport to enter an 'invalid state', which prevents them from doing anything in-passport, and provides a button that deletes their passport from local storage.

```
[PRETIX] Deleting {"email":"ivansergchub+visitor@gmail.com","order_id":"UDESF","name":"Ivan Chub","role":"visitor","residence":"","email_token":"919626","visitor_date_ranges":[{"date_to":"2023-05-06T21:59:59Z","date_from":"2023-04-29T22:00:00Z"},{"date_to":"2023-04-29T21:59:59Z","date_from":"2023-04-22T22:00:00Z"},{"date_to":"2023-04-22T21:59:59Z","date_from":"2023-04-15T22:00:00Z"}]}
```

- I reactivate (basically the same thing as adding a new order) in pretix for my test visitor. This causes a pretix participant to be created in the zupass database, which is the correct thing. This allows me to sign in, and my passport is immediately a visitor witht the correct dates displayed.


```
[PRETIX] Inserting {"role":"visitor","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"W9XWV","email_token":"","visitor_date_ranges":[{"date_from":"2023-04-15T22:00:00Z","date_to":"2023-04-22T21:59:59Z"}]}
```

- I reactivate the other order, which has two tickets. This causes the preti participant to be updated in the zupass database, and gets loaded into the passport.

```
[PRETIX] Updating {"email":"ivansergchub+visitor@gmail.com","order_id":"W9XWV","name":"Ivan Chub","role":"visitor","residence":"","email_token":"092224","visitor_date_ranges":[{"date_to":"2023-04-22T21:59:59Z","date_from":"2023-04-15T22:00:00Z"}]} to {"role":"visitor","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"UDESF","email_token":"","visitor_date_ranges":[{"date_from":"2023-04-29T22:00:00Z","date_to":"2023-05-06T21:59:59Z"},{"date_from":"2023-04-22T22:00:00Z","date_to":"2023-04-29T21:59:59Z"},{"date_from":"2023-04-15T22:00:00Z","date_to":"2023-04-22T21:59:59Z"}]
```

- In the order which has multiple visitor tickets (positions), I cancel one of them. My expected behavior is that the other ticket remains synced to the participant in zupass. The thing that occurs is correct!

```
[PRETIX] Updating {"email":"ivansergchub+visitor@gmail.com","order_id":"W9XWV","name":"Ivan Chub","role":"visitor","residence":"","email_token":"092224","visitor_date_ranges":[{"date_to":"2023-05-06T21:59:59Z","date_from":"2023-04-29T22:00:00Z"},{"date_to":"2023-04-29T21:59:59Z","date_from":"2023-04-22T22:00:00Z"},{"date_to":"2023-04-22T21:59:59Z","date_from":"2023-04-15T22:00:00Z"}]} to {"role":"visitor","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"UDESF","email_token":"","visitor_date_ranges":[{"date_from":"2023-04-22T22:00:00Z","date_to":"2023-04-29T21:59:59Z"},{"date_from":"2023-04-15T22:00:00Z","date_to":"2023-04-22T21:59:59Z"}]}
```

- I readd the ticket which i just deleted, and it gets properly re-added to the pretix participant.

```
[PRETIX] Updating {"email":"ivansergchub+visitor@gmail.com","order_id":"W9XWV","name":"Ivan Chub","role":"visitor","residence":"","email_token":"092224","visitor_date_ranges":[{"date_to":"2023-04-29T21:59:59Z","date_from":"2023-04-22T22:00:00Z"},{"date_to":"2023-04-22T21:59:59Z","date_from":"2023-04-15T22:00:00Z"}]} to {"role":"visitor","email":"ivansergchub+visitor@gmail.com","name":"Ivan Chub","residence":"","order_id":"UDESF","email_token":"","visitor_date_ranges":[{"date_from":"2023-04-22T22:00:00Z","date_to":"2023-04-29T21:59:59Z"},{"date_from":"2023-04-29T22:00:00Z","date_to":"2023-05-06T21:59:59Z"},{"date_from":"2023-04-15T22:00:00Z","date_to":"2023-04-22T21:59:59Z"}]}
```

ichub commented on apps/passport-server/src/apis/pretixAPI.ts:
To clean up the `pretixSync.ts` file, I moved functions that deal with the API directly to this file.

ichub commented on packages/passport-interface/src/zuzalu.ts:
```diff
@@ -8,19 +8,23 @@ export interface ZuParticipant {
-export interface BackendUser {
-  identifier: string;
-  status: number;
-  createdAt: Date;
-  encryptedBlob: string;
-  updatedAt: Date;
+export interface DateRange {
+  date_from?: string | null;
+  date_to?: string | null;
 }
 
-export interface ClientUser extends BackendUser {
-  masterKey: string;
-  serverPassword: string;
```
these `BackendUser` and `ClientUser` interfaces were never actually used.

dcposch commented on apps/passport-client/components/shared/PCDCard.tsx:
```diff
@@ -24,21 +24,40 @@ export function PCDCard({
-    ? "VERIFIED ZUZALU PASSPORT"
-    : displayOptions?.header?.toUpperCase() ?? "PCD";
+
+  const visitorStatus = getVisitorStatus(state.self);
```
all this visitor stuff only applies if `isZuzaluIdentity`, right? like, it shouldn't affect the ethereum PCD etc

ichub replied:
Yes, the fact that i had to made a change here is not ideal.

dcposch commented on apps/passport-client/components/shared/PCDCard.tsx:
```diff
@@ -24,21 +24,40 @@ export function PCDCard({
+
+  let header = displayOptions?.header?.toUpperCase() ?? "PCD";
+
+  if (isZuzaluIdentity) {
```
i'd consider `let expired = false` above, then

```
if (isZuzaluIdentity) {
  expired = isZuzaluIDExpired(...);
  header = expired ? "EXPIRED" : "VERIFIED ...";
}
```
that pulls a bit of zuzalu-specific logic outside of the `PCDCard` render function & ensures it only runs if `isZuzaluID`

dcposch commented on apps/passport-client/components/shared/ZuzaluCard.tsx:
```diff
@@ -33,24 +44,60 @@ export function ZuzaluCardBody({
         <H3 col="var(--primary-dark)">{name}</H3>
         <InfoLine>{email}</InfoLine>
         <InfoLine>{residence}</InfoLine>
+        <VisitorDateSection participant={actualParticipant} />
```
👍

dcposch commented on apps/passport-client/src/participant.ts:
```diff
@@ -27,3 +32,66 @@ export async function pollParticipant(
+}
+
+const ZUZALU_START_DATE = "2023-03-24";
+const ZUZALU_END_DATE = "2023-05-26";
```
hardcoding this into the UI (as opposed to say in `pretixsync`) feels dirty

dcposch commented on apps/passport-server/src/apis/pretixAPI.ts:
```diff
@@ -0,0 +1,110 @@
+import { requireEnv } from "../util/util";
```
pulling into this file makes sense

dcposch commented on apps/passport-server/migrations/10_visitor_dates.sql:
```diff
@@ -0,0 +1,2 @@
+alter table pretix_participants
+add column visitor_date_ranges JsonB;
```
ah i see, `visitor_date_ranges` comes from the SQL column name. reasonable

dcposch reviewed (approved):
LGTM

left comments

if possible i'd calculate date ranges on the server (so every paticipant has, say, `dateRanges`; for resident/organizer it's a single range that spans the whole event, for visitors it's a list of ranges that comes from `visitor_date_ranges` in the DB)

this might let you simplify UI logic a bit

dcposch commented:
@ichub if i undestand right, this PR doesn't modify the four semaphore groups, right?

so the `VISITOR` timeboxing is just for the QR + QR verification

& all zuzalu participiants can still log in to the website indefinitely (that seems good)

dcposch commented:
## Important - who will be deleted?

before we ship this, let's figure out the list of current passport holders that will be deleted.

ichub commented:
> @ichub if i undestand right, this PR doesn't modify the four semaphore groups, right?

It modifies the definition of who counts as a visitor, and also removes users who are no longer in pretix, so it does modify the semaphore groups in that sense. On the other hand, it doesn't create new semaphore groups, nor does it remove old ones.

> so the `VISITOR` timeboxing is just for the QR + QR verification

Visitor timeboxing is just for in-person access, correct. Visitors with 'expired' passes are still able to log in to zuzalu.city, etc.

> & all zuzalu participiants can still log in to the website indefinitely (that seems good)

All zuzalu participants which have not been removed from pretix, by cancelling their order, for example, yes.

ichub commented:
> ## Important - who will be deleted?
> 
> before we ship this, let's figure out the list of current passport holders that will be deleted.

Good idea.

gubsheep commented:
The expected behaviors you describe in the PR body SGTM. One question: can you include a screenshot for the state described below?

| user had several visitor entries in pretix, and I deleted all of them. I expect this to cause the pretix user to be deleted from the database. the behavior I saw matched my expectations. On the client side, this caused the user's passport to enter an 'invalid state', which prevents them from doing anything in-passport, and provides a button that deletes their passport from local storage.

Can you also confirm that you have the same behavior when the user has a resident entry in pretix, and you delete their resident entry?

ichub commented:
@dcposch - to find out who is deleted, I cloned the production `pretix_participants` table into my local development environment, ran the pretix sync, and saved a json file of the pretix participants who got deleted. I then ran this script on them:

```js
const fs = require("fs");

const removed = JSON.parse(fs.readFileSync("./removed.json").toString());
const totalVisitors = removed.filter((p) => p.role === "visitor");
const totalResidents = removed.filter((p) => p.role === "resident");
const totalOrganizers = removed.filter((p) => p.role === "organizer");
const totalWithoutEmailCode = removed.filter((p) => p.email_token === null);

console.log(`removed pretix participants: `, removed.length);
console.log(`removed visitors `, totalVisitors.length);
console.log(`removed residents `, totalResidents.length);
console.log(`removed organizers `, totalOrganizers.length);
console.log(`never logged in `, totalWithoutEmailCode.length);
```

and I got the following output:

```
removed pretix participants:  41
removed visitors  39
removed residents  2
removed organizers  0
never logged in  9
```

I also shared that json file in discord for you to check out, if you want.

My take is that this deletion seems to be what we want.

ichub commented:
> The expected behaviors you describe in the PR body SGTM. One question: can you include a screenshot for the state described below?
> 
> | user had several visitor entries in pretix, and I deleted all of them. I expect this to cause the pretix user to be deleted from the database. the behavior I saw matched my expectations. On the client side, this caused the user's passport to enter an 'invalid state', which prevents them from doing anything in-passport, and provides a button that deletes their passport from local storage.

<img width="861" alt="Screenshot 2023-04-30 at 7 23 15 PM" src="https://user-images.githubusercontent.com/2636237/235367343-d9ba785b-3a51-419a-bba3-64c6aafd906c.png">
<img width="617" alt="Screenshot 2023-04-30 at 7 23 39 PM" src="https://user-images.githubusercontent.com/2636237/235367349-d3e5c6c7-edaa-4ed2-bc20-36944b0a52e6.png">


> Can you also confirm that you have the same behavior when the user has a resident entry in pretix, and you delete their resident entry?

<img width="860" alt="Screenshot 2023-04-30 at 7 25 21 PM" src="https://user-images.githubusercontent.com/2636237/235367354-febd7c85-9c57-4f12-ba9a-1813666c0575.png">
<img width="863" alt="Screenshot 2023-04-30 at 7 27 04 PM" src="https://user-images.githubusercontent.com/2636237/235367358-d186bb1c-5333-407f-b1b4-840a32ba1de0.png">
<img width="565" alt="Screenshot 2023-04-30 at 7 27 40 PM" src="https://user-images.githubusercontent.com/2636237/235367360-6132fed3-6152-4345-b824-3b856dc69f33.png">
